pandas/core/generic.py

```diff
@@ -4506,22 +4506,104 @@ def astype(self, dtype, copy=True, errors='raise', **kwargs):
 
     def copy(self, deep=True):
         """
-        Make a copy of this objects data.
+        Make a copy of this object's indices and data.
+
+        When `deep=True` (default), a new object will be created with a
+        copy of the calling object's data and indices. Modifications to
+        the data or indices of the copy will not be reflected in the
+        original object (see notes below).
+
+        When `deep=False`, a new object will be created without copying
+        the calling object's data or index (only references to the data
+        and index are copied). Any changes to the data of the original
+        will be reflected in the shallow copy (and vice versa). Changes
+        to the index will result in a new index as indices are immutable.
```

Comment:
I think this line "Changes to the index will result in a new index as indices are immutable" is a bit out of place here, as this is not related to `deep=False`

For the fact that indices are immutable, I would add a short explanation about that in the Notes section (it is a bit too detailed for the extended summary at the beginning of the docstring). 
The main thing is that, even for `deep=True`, the index *data* (the underlying numpy array) is not copied, only the object (the Index class holding the array data). This is done for performance reasons: because Index is immutable, it is not needed to copy the data.
(see https://github.com/pandas-dev/pandas/issues/19862 for discussion on it)